src/state.rs

```diff
@@ -0,0 +1,585 @@
+use sctk::reexports::client::protocol::wl_pointer::WlPointer;
+use sctk::reexports::client::protocol::wl_seat::WlSeat;
+use sctk::reexports::client::{Connection, QueueHandle, Dispatch, Proxy};
+use sctk::reexports::protocols::wp::primary_selection::zv1::client::{
+    zwp_primary_selection_device_v1::ZwpPrimarySelectionDeviceV1,
+    zwp_primary_selection_source_v1::ZwpPrimarySelectionSourceV1,
+};
+use sctk::reexports::calloop::{LoopHandle, PostAction};
+use wayland_backend::client::ObjectId;
+
+use crate::mime::{normalize_to_lf, MimeType, ALLOWED_MIME_TYPES};
+
+pub struct State {
+    pub primary_selection_manager_state: Option<PrimarySelectionManagerState>,
+    pub data_device_manager_state: Option<DataDeviceManagerState>,
+    pub reply_tx: Sender<Result<String>>,
+    pub exit: bool,
+
+    registry_state: RegistryState,
+    seat_state: SeatState,
+
+    seats: HashMap<ObjectId, ClipboardSeatState>,
+    /// The latest seat which got an event.
+    latest_seat: Option<ObjectId>,
+
+    loop_handle: LoopHandle<'static, Self>,
+    queue_handle: QueueHandle<Self>,
+
+    primary_sources: Vec<PrimarySelectionSource>,
+    primary_selection_content: Rc<[u8]>,
+
+    data_sources: Vec<CopyPasteSource>,
+    data_selection_content: Rc<[u8]>,
+}
+
+impl State {
+    #[must_use]
+    pub fn new(
+        globals: &GlobalList,
+        queue_handle: &QueueHandle<Self>,
+        loop_handle: LoopHandle<'static, Self>,
+        reply_tx: Sender<Result<String>>,
+    ) -> Option<Self> {
+        let mut seats = HashMap::new();
+
+        let data_device_manager_state = DataDeviceManagerState::bind(globals, queue_handle).ok();
+        let primary_selection_manager_state =
+            PrimarySelectionManagerState::bind(globals, queue_handle).ok();
+
+        // When both globals are not available nothing could be done.
+        if data_device_manager_state.is_none() && primary_selection_manager_state.is_none() {
+            return None;
+        }
+
+        let seat_state = SeatState::new(globals, queue_handle);
+        for seat in seat_state.seats() {
+            seats.insert(seat.id(), Default::default());
+        }
+
+        Some(Self {
+            registry_state: RegistryState::new(globals),
+            primary_selection_content: Rc::from([]),
+            data_selection_content: Rc::from([]),
+            queue_handle: queue_handle.clone(),
+            primary_selection_manager_state,
+            primary_sources: Vec::new(),
+            data_device_manager_state,
+            data_sources: Vec::new(),
+            latest_seat: None,
+            loop_handle,
+            exit: false,
+            seat_state,
+            reply_tx,
+            seats,
+        })
+    }
+
+    /// Store selection for the given target.
+    ///
+    /// Selection source is only created when `Some(())` is returned.
+    pub fn store_selection(&mut self, ty: SelectionTarget, contents: String) -> Option<()> {
```

Comment:
Should this return a Result similar to `load_selection`?

Reply:
The reason for that is that `load` wants a reply from the clipboard, so if we don't propagate the result and reply with `error`, we could hang the client, however the `store` doesn't need anything like that.

The only use case for store is to `log` something, which wasn't the case before.

I could document why one has and other doesn't return a result.

The reason I use `Option` is to simply rely on `?` to make the code less verbose.

Reply:
Ok, I figured it just makes it less verbose. I think it's fine as is, but just wanted to check